components/tidb_query_datatype/src/codec/row/v2/encoder_for_test.rs

```diff
@@ -89,8 +90,100 @@ impl Column {
     }
 }
 
+/// Checksum
+/// - HEADER(1 byte)
+///   - VER: version(3 bit)
+///   - E:   has extra checksum
+/// - CHECKSUM(4 bytes)
+///   - little-endian CRC32(IEEE) when hdr.ver = 0 (default)
+pub trait CheckSumHandler {
+    // update_col updates the checksum with the encoded value of the column.
+    fn checksum(&mut self, buf: &[u8]) -> Result<()>;
+
+    // header_value returns the checksum header value.
+    fn header_value(&self) -> u8;
+
+    // value returns the checksum value.
+    fn value(&self) -> u32;
+}
+
+pub struct Crc32RowChecksumHandler {
+    header: ChecksumHeader,
+    hasher: crc32fast::Hasher,
+}
+
+impl CheckSumHandler for Crc32RowChecksumHandler {
+    fn checksum(&mut self, buf: &[u8]) -> Result<()> {
+        self.hasher.update(buf);
+        Ok(())
+    }
+
+    fn header_value(&self) -> u8 {
+        self.header.value()
+    }
+
+    fn value(&self) -> u32 {
+        self.hasher.clone().finalize()
+    }
+}
+
+pub struct ChecksumHeader(u8);
+
+impl ChecksumHeader {
+    fn new() -> Self {
+        ChecksumHeader(0)
+    }
+
+    #[cfg(test)]
+    fn set_version(&mut self, ver: u8) {
+        self.0 = ver & 0b0000_0111;
+    }
+
+    #[cfg(test)]
+    fn set_extra_checksum(&mut self) {
+        self.0 |= 0b0000_1000;
+    }
+
+    fn value(&self) -> u8 {
+        self.0
+    }
+}
+
+impl Crc32RowChecksumHandler {
+    pub fn new() -> Self {
+        Crc32RowChecksumHandler {
+            header: ChecksumHeader::new(),
+            hasher: crc32fast::Hasher::new(),
+        }
+    }
+}
+
+impl Default for Crc32RowChecksumHandler {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
 pub trait RowEncoder: NumberEncoder {
     fn write_row(&mut self, ctx: &mut EvalContext, columns: Vec<Column>) -> Result<()> {
+        self.write_row_impl(ctx, columns, None)
+    }
+
+    fn write_row_with_checksum(
+        &mut self,
+        ctx: &mut EvalContext,
+        columns: Vec<Column>,
+        handler: Option<&mut dyn CheckSumHandler>,
+    ) -> Result<()> {
+        self.write_row_impl(ctx, columns, handler)
+    }
+
+    fn write_row_impl(
+        &mut self,
+        ctx: &mut EvalContext,
+        columns: Vec<Column>,
+        mut checksum_handler: Option<&mut dyn CheckSumHandler>,
+    ) -> Result<()> {
         let mut is_big = false;
```

Comment:
Do we need to add a flag like `has_checksum`?

Reply:
It would be added later in the `RowSlice` struct which is actually used decoder.